### ISSUE TITLE
How to calculate the total frame count for gif?

### ISSUE THREAD

**Issue by lemarian** [rs/SDWebImage#2158]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / tvos
 Platform Version        | e.g. 8.0
 SDWebImage Version      | e.g. 3.7.6
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 7.3
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Comment by dreampiggy** [comment]
Are you sure you're using 3.7.6 version? This may return a `UIAnimatedImage` and you can check `-[UIImage images]` array to get the count.

**Comment by lemarian** [comment]
I am using version 3.8.2

**Comment by lemarian** [comment]
self.imgView.sd_setImage(with: self.urlImage!, placeholderImage: nil, options: [SDWebImageOptions.continueInBackground, SDWebImageOptions.lowPriority, SDWebImageOptions.refreshCached, SDWebImageOptions.handleCookies, SDWebImageOptions.retryFailed,.transformAnimatedImage], completed:  {(image, error, cacheType, url) in
            
            
            if error == nil && image != nil  {
                
                self.acIn.isHidden = true
                self.acIn.stopAnimating()
}
})
This is my block of code

**Comment by dreampiggy** [comment]
Just check `image.images.count != 0` and you're done.

**Comment by lemarian** [comment]
I am loading gif in a imageview and I want to count the total number of frame in a gif to calculate the total duration of gif, How will image.images.count resolve what i want to attain?

**Comment by dreampiggy** [comment]
@lemarian See there API docs:
https://developer.apple.com/documentation/uikit/uiimage/1624149-animatedimagewithimages
https://developer.apple.com/documentation/uikit/uiimage/1624155-duration
https://developer.apple.com/documentation/uikit/uiimage/1624117-images

**Comment by lemarian** [comment]
This still doesn't solve my problem. I want to calculate the total duration of the gif. The total duration  of gif playing.

**Comment by lemarian** [comment]
I want to calculate the total duration a particular gif plays.. as soon as it finishes playing, I have to switch to another view. So i need to calculate the exact time frame of the gif

**Comment by dreampiggy** [comment]
Don't really catch up with your custom use case. You can try to decoding the image with raw Data and do this. Or to find someother specify imageView such as FLAnimatedImageView to do something after one loop finished.

Our SDWebImage 5.x introduce a like SDAnimatedImageView and provide a property 'currentLoopCount' which you can use KVO on it. But this need time to release. So maybe you should try the other solution.

**Comment by dreampiggy** [comment]
@lemarian Well...Seems I'm not catup up with the idea at that time.

If you need to `switch to another view after the current gif play finished`. First solution, you setup a `NSTimer` or something to record the time. Then use `UIImage.duration` to get the total duration the image will play in each loop. So now you can switch to another view after the total duration time passed. The disadvantage of this solution is that the timer must be accurate or there may be some synchonized issue.

The second solution, using a special imageView which support a callback after animated image play finished. For `UIImageView` with `UIAnimatedImage`, use KVO to observe the `animating` property. When that change to false, switch to another view. Attention this assume that the image loop count to be 1(Or the image keep animating until final loop finished)

For `FLAnimatedImageView`, use `loopCompletionBlock`.
For `SDAnimatedImageView` from our 5.x branch, use KVO on `animating`(Means all loop finished and image stop animating) or `currentLoopCount` (Means current loop finished and will begin next loop)
